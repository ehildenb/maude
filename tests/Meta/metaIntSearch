#!/bin/sh

MAUDE_LIB=$srcdir/../../src/Main
export MAUDE_LIB

../../src/Main/maude \
  < $srcdir/metaIntSearch.maude -no-banner -no-advise \
  > metaIntSearch.out 2>&1

diff $srcdir/metaIntSearch.expected metaIntSearch.out > /dev/null 2>&1
