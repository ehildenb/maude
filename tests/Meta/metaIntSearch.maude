set show timing off .
set show advisories off .

***
*** Test search in the meta-interpreter without and with path.
***

load metaInterpreter

mod SEARCH is
  sort State .
  ops a b c d e : -> State .

  rl [a-move1] : a => b .
  rl [a-move2] : a => c .
  rl [b-move] : b => d .
  rl [c-move] : c => e .
endm

mod SEARCH-TEST is
  pr META-INTERPRETER .

  op me : -> Oid .
  op User : -> Cid .
  op soln:_ : Nat -> Attribute .

  vars X Y Z : Oid .
  var AS : AttributeSet .
  var N : Nat .

  rl < X : User | AS > createdInterpreter(X, Y, Z) =>
     < X : User | AS > insertModule(Z, X, upModule('SEARCH, true)) .

  rl < X : User | soln: N, AS > insertedModule(X, Y) =>
     < X : User | AS > getSearchResult(Y, X, 'SEARCH, 'a.State, 'X:State, nil, '*, 20, N) .
endm

erew in SEARCH-TEST : <> 
< me : User | soln: 0 > 
createInterpreter(interpreterManager, me, none) .

erew in SEARCH-TEST : <> 
< me : User | soln: 1 > 
createInterpreter(interpreterManager, me, none) .

erew in SEARCH-TEST : <> 
< me : User | soln: 2 > 
createInterpreter(interpreterManager, me, none) .

erew in SEARCH-TEST : <> 
< me : User | soln: 3 > 
createInterpreter(interpreterManager, me, none) .

erew in SEARCH-TEST : <> 
< me : User | soln: 4 > 
createInterpreter(interpreterManager, me, none) .

erew in SEARCH-TEST : <> 
< me : User | soln: 5 > 
createInterpreter(interpreterManager, me, none) .

mod SEARCH-CACHE-TEST is
  pr META-INTERPRETER .

  op me : -> Oid .
  op User : -> Cid .
  op soln:_ : Nat -> Attribute .
  op got:_  : Msg -> Attribute .

  vars X Y Z : Oid .
  var AS : AttributeSet .
  vars N M : Nat .
  var S : Substitution .
  var T : Term .
  var TY : Type .

  rl < X : User | AS > createdInterpreter(X, Y, Z) =>
     < X : User | AS > insertModule(Z, X, upModule('SEARCH, true)) .

  rl < X : User | soln: N, AS > insertedModule(X, Y) =>
     < X : User | AS, soln: 1 > getSearchResult(Y, X, 'SEARCH, 'a.State, 'X:State, nil, '*, 20, 0) .

  rl < X : User | soln: N, AS > gotSearchResult(X, Y, M, T, TY, S) =>
     < X : User | AS, soln: (N + 1), got: gotSearchResult(X, Y, M, T, TY, S) >
     getSearchResult(Y, X, 'SEARCH, 'a.State, 'X:State, nil, '*, 20, N) .
endm

erew in SEARCH-CACHE-TEST : <> 
< me : User | soln: 0 > 
createInterpreter(interpreterManager, me, none) .


mod SEARCH-PATH-TEST is
  pr META-INTERPRETER .

  op me : -> Oid .
  op User : -> Cid .
  op soln:_ : Nat -> Attribute .

  vars X Y Z : Oid .
  var AS : AttributeSet .
  var N : Nat .

  rl < X : User | AS > createdInterpreter(X, Y, Z) =>
     < X : User | AS > insertModule(Z, X, upModule('SEARCH, true)) .

  rl < X : User | soln: N, AS > insertedModule(X, Y) =>
     < X : User | AS > getSearchResultAndPath(Y, X, 'SEARCH, 'a.State, 'X:State, nil, '*, 20, N) .
endm

erew in SEARCH-PATH-TEST : <> 
< me : User | soln: 0 > 
createInterpreter(interpreterManager, me, none) .

erew in SEARCH-PATH-TEST : <> 
< me : User | soln: 1 > 
createInterpreter(interpreterManager, me, none) .

erew in SEARCH-PATH-TEST : <> 
< me : User | soln: 2 > 
createInterpreter(interpreterManager, me, none) .

erew in SEARCH-PATH-TEST : <> 
< me : User | soln: 3 > 
createInterpreter(interpreterManager, me, none) .

erew in SEARCH-PATH-TEST : <> 
< me : User | soln: 4 > 
createInterpreter(interpreterManager, me, none) .

erew in SEARCH-PATH-TEST : <> 
< me : User | soln: 5 > 
createInterpreter(interpreterManager, me, none) .

mod SEARCH-PATH-CACHE-TEST is
  pr META-INTERPRETER .

  op me : -> Oid .
  op User : -> Cid .
  op soln:_ : Nat -> Attribute .
  op got:_  : Msg -> Attribute .

  vars X Y Z : Oid .
  var AS : AttributeSet .
  vars N M : Nat .
  var S : Substitution .
  var T : Term .
  var TY : Type .
  var TR : Trace .

  rl < X : User | AS > createdInterpreter(X, Y, Z) =>
     < X : User | AS > insertModule(Z, X, upModule('SEARCH, true)) .

  rl < X : User | soln: N, AS > insertedModule(X, Y) =>
     < X : User | AS, soln: 1 > getSearchResultAndPath(Y, X, 'SEARCH, 'a.State, 'X:State, nil, '*, 20, 0) .

  rl < X : User | soln: N, AS > gotSearchResultAndPath(X, Y, M, T, TY, S, TR) =>
     < X : User | AS, soln: (N + 1), got: gotSearchResultAndPath(X, Y, M, T, TY, S, TR) >
     getSearchResultAndPath(Y, X, 'SEARCH, 'a.State, 'X:State, nil, '*, 20, N) .
endm

erew in SEARCH-PATH-CACHE-TEST : <> 
< me : User | soln: 0 > 
createInterpreter(interpreterManager, me, none) .
