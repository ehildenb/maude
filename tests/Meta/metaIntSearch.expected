==========================================
erewrite in SEARCH-TEST : <> < me : User | soln: 0 > createInterpreter(
    interpreterManager, me, none) .
rewrites: 3
result Configuration: <> < me : User | none > gotSearchResult(me, interpreter(
    0), 0, 'a.State, 'State, 
  'X:State <- 'a.State)
==========================================
erewrite in SEARCH-TEST : <> < me : User | soln: 1 > createInterpreter(
    interpreterManager, me, none) .
rewrites: 4
result Configuration: <> < me : User | none > gotSearchResult(me, interpreter(
    0), 1, 'b.State, 'State, 
  'X:State <- 'b.State)
==========================================
erewrite in SEARCH-TEST : <> < me : User | soln: 2 > createInterpreter(
    interpreterManager, me, none) .
rewrites: 5
result Configuration: <> < me : User | none > gotSearchResult(me, interpreter(
    0), 2, 'c.State, 'State, 
  'X:State <- 'c.State)
==========================================
erewrite in SEARCH-TEST : <> < me : User | soln: 3 > createInterpreter(
    interpreterManager, me, none) .
rewrites: 6
result Configuration: <> < me : User | none > gotSearchResult(me, interpreter(
    0), 3, 'd.State, 'State, 
  'X:State <- 'd.State)
==========================================
erewrite in SEARCH-TEST : <> < me : User | soln: 4 > createInterpreter(
    interpreterManager, me, none) .
rewrites: 7
result Configuration: <> < me : User | none > gotSearchResult(me, interpreter(
    0), 4, 'e.State, 'State, 
  'X:State <- 'e.State)
==========================================
erewrite in SEARCH-TEST : <> < me : User | soln: 5 > createInterpreter(
    interpreterManager, me, none) .
rewrites: 7
result Configuration: <> < me : User | none > noSuchResult(me, interpreter(0),
    4)
==========================================
erewrite in SEARCH-CACHE-TEST : <> < me : User | soln: 0 > createInterpreter(
    interpreterManager, me, none) .
rewrites: 17
result Configuration: <> < me : User | soln: 6,got: gotSearchResult(me,
    interpreter(0), 0, 'a.State, 'State, 
  'X:State <- 'a.State),got: gotSearchResult(me, interpreter(0), 1, 'b.State,
    'State, 
  'X:State <- 'b.State),got: gotSearchResult(me, interpreter(0), 1, 'c.State,
    'State, 
  'X:State <- 'c.State),got: gotSearchResult(me, interpreter(0), 1, 'd.State,
    'State, 
  'X:State <- 'd.State),got: gotSearchResult(me, interpreter(0), 1, 'e.State,
    'State, 
  'X:State <- 'e.State) > noSuchResult(me, interpreter(0), 0)
==========================================
erewrite in SEARCH-PATH-TEST : <> < me : User | soln: 0 > createInterpreter(
    interpreterManager, me, none) .
rewrites: 3
result Configuration: <> < me : User | none > gotSearchResultAndPath(me,
    interpreter(0), 0, 'a.State, 'State, 
  'X:State <- 'a.State, nil)
==========================================
erewrite in SEARCH-PATH-TEST : <> < me : User | soln: 1 > createInterpreter(
    interpreterManager, me, none) .
rewrites: 4
result Configuration: <> < me : User | none > gotSearchResultAndPath(me,
    interpreter(0), 1, 'b.State, 'State, 
  'X:State <- 'b.State, {'a.State,'State,rl 'a.State => 'b.State [label(
    'a-move1)] .})
==========================================
erewrite in SEARCH-PATH-TEST : <> < me : User | soln: 2 > createInterpreter(
    interpreterManager, me, none) .
rewrites: 5
result Configuration: <> < me : User | none > gotSearchResultAndPath(me,
    interpreter(0), 2, 'c.State, 'State, 
  'X:State <- 'c.State, {'a.State,'State,rl 'a.State => 'c.State [label(
    'a-move2)] .})
==========================================
erewrite in SEARCH-PATH-TEST : <> < me : User | soln: 3 > createInterpreter(
    interpreterManager, me, none) .
rewrites: 6
result Configuration: <> < me : User | none > gotSearchResultAndPath(me,
    interpreter(0), 3, 'd.State, 'State, 
  'X:State <- 'd.State, {'a.State,'State,rl 'a.State => 'b.State [label(
    'a-move1)] .}
{'b.State,'State,rl 'b.State => 'd.State [label('b-move)] .})
==========================================
erewrite in SEARCH-PATH-TEST : <> < me : User | soln: 4 > createInterpreter(
    interpreterManager, me, none) .
rewrites: 7
result Configuration: <> < me : User | none > gotSearchResultAndPath(me,
    interpreter(0), 4, 'e.State, 'State, 
  'X:State <- 'e.State, {'a.State,'State,rl 'a.State => 'c.State [label(
    'a-move2)] .}
{'c.State,'State,rl 'c.State => 'e.State [label('c-move)] .})
==========================================
erewrite in SEARCH-PATH-TEST : <> < me : User | soln: 5 > createInterpreter(
    interpreterManager, me, none) .
rewrites: 7
result Configuration: <> < me : User | none > noSuchResult(me, interpreter(0),
    4)
==========================================
erewrite in SEARCH-PATH-CACHE-TEST : <> < me : User | soln: 0 >
    createInterpreter(interpreterManager, me, none) .
rewrites: 17
result Configuration: <> < me : User | soln: 6,got: gotSearchResultAndPath(me,
    interpreter(0), 0, 'a.State, 'State, 
  'X:State <- 'a.State, nil),got: gotSearchResultAndPath(me, interpreter(0), 1,
    'b.State, 'State, 
  'X:State <- 'b.State, {'a.State,'State,rl 'a.State => 'b.State [label(
    'a-move1)] .}),got: gotSearchResultAndPath(me, interpreter(0), 1, 'c.State,
    'State, 
  'X:State <- 'c.State, {'a.State,'State,rl 'a.State => 'c.State [label(
    'a-move2)] .}),got: gotSearchResultAndPath(me, interpreter(0), 1, 'd.State,
    'State, 
  'X:State <- 'd.State, {'a.State,'State,rl 'a.State => 'b.State [label(
    'a-move1)] .}
{'b.State,'State,rl 'b.State => 'd.State [label('b-move)] .}),got:
    gotSearchResultAndPath(me, interpreter(0), 1, 'e.State, 'State, 
  'X:State <- 'e.State, {'a.State,'State,rl 'a.State => 'c.State [label(
    'a-move2)] .}
{'c.State,'State,rl 'c.State => 'e.State [label('c-move)] .}) > noSuchResult(
    me, interpreter(0), 0)
Bye.
